web3/_utils/abi.py

```diff
@@ -342,6 +347,66 @@ def merge_args_and_kwargs(function_abi, args, kwargs):
         return tuple()
 
 
+TUPLE_PREFIX = 'tuple'
+TUPLE_PREFIX_LEN = len(TUPLE_PREFIX)
+
+
+def _convert_abi_input(comp, arg):
+    """
+    Converts an argument ``arg`` corresponding to an ABI component ``comp``
+    into a plain value (for non-tuple components) or a properly converted and
+    ordered sequence (for tuple list components or tuple components).
+    """
+    abi_type_str = comp['type']
+
+    if not abi_type_str.startswith(TUPLE_PREFIX):
+        # Component is non-tuple.  Just return value.
+        return arg
+
+    # Component is tuple (possibly tuple list)...
+
+    array_dims = abi_type_str[TUPLE_PREFIX_LEN:]
+    if array_dims:
+        # Component is tuple list (i.e. it has array dimensions).  Construct
+        # new ABI component without array dimensions and recurse with it for
+        # each element in arg.
+        new_comp = copy.copy(comp)
+        new_comp['type'] = TUPLE_PREFIX
+
+        # Assume `arg` is iterable
+        return type(arg)(_convert_abi_input(new_comp, a) for a in arg)
+
+    # Component is non-list tuple...
+
+    sub_comps = comp['components']
+
+    if isinstance(arg, abc.Mapping):
+        # Arg is mapping.  Convert to properly ordered sequence.
+        arg = tuple(arg[c['name']] for c in sub_comps)
```

Comment:
It's going to take me a long time to not have a knee-jerk reflex that python dictionaries aren't ordered (since they now are)

Reply:
If you meant `sub_comps`, I believe that should always be a `list` instance.